Issue37

tarwirdur opened the pull request:
https://github.com/wladich/nakarte/issues/37

tarwirdur commented:
Впрочем там из-за `sql.js` неприличное увеличение `vendor.js`:

```
File sizes after gzip:

  825.81 KB (+687.71 KB)  build/static/js/vendor.53d8d5ae.js
  292.61 KB               build/static/js/mapillary.e7e6f78e.chunk.js
  98.32 KB (+1.11 KB)     build/static/js/main.d7451ac2.js
  35.87 KB (+44 B)        build/static/css/main.6099dd31.css
  6.08 KB                 build/static/css/vendor.86456fe6.css
```

Пока не разобрался как сделана асинхронная загрузка для mapillary.

tarwirdur commented:
>2. Т.е. надо придумать, как сделать не сильно хуже, чем сейчас -- может быть можно сделать иконку на эту тему.

Иконки искал, но подходящих не увидел.
Поищу ещё.

>Красным цветом выделяются строки меню, которые могут привести к созданию невалидного JNX (> 50000 тайлов). Для Rmap это не актуально.

Насколько я понимаю, все данные хранятся при экспорте в оперативной памяти. И 50 000 тайлов по оценке - 1gb. Выглядит как предупреждение о большом расходовании памяти.

>5. Выбор формата должен сохраняться в hashState

Спорно. Если я часто пользуюсь nakarte.tk и osmand - и мне пришлет владелец garmin ссылку на nakarte - экспорт будет предложен в jnx даже если я до этого всегда экспортировал в osmand.

>6. Во всех контекстных меню (например, для треков и для страниц в контроле печати) появились пункты "undefined".

o_O. else пропустил. :(

Но, учитывая п.2 надо убирать selectmenu из contextmenu. (.action {cursor: pointer;} оставлю).

wladich commented:
2. Может придётся что-то нарисовать или скомпилировать -- по моему мнению, при свёрнутом контроле (т.е. когда видна только кнопка) должно быть примерно понятно, что есть экспорт для и навигатора, и для телефона.

4. 50000 это ограничение формата jnx. В нормальном диалоге можно будет показывать полноценное предупреждение.

5. Ссылка (с форматом экспорта) должна работать аналогично ссылке для PDF: если нет выбранного прямоугольникa, то в ссылке совсем нет соотв. параметра. Т.е. если пришлют ссылку просто на место, то формат не поменяется. А вот если пришлёт ссылку на JNX с просьбой скачать именно JNX, то всё сработает как надо. И главное -- при F5 состояние интерфейса должно сохраняться. Если очень хочется, то последний использованный формат можно дополнительно хранить в localStorage

tarwirdur commented:
@wladich 
>    Большой размер библиотеки без загрузки по требованию

fixed

>    Выбор формата должен сохраняться в hashState

fixed

>    Во всех контекстных меню (например, для треков и для страниц в контроле печати) появились пункты "undefined".

fixed

>    Замена надписи на Export. Сейчас новый пользователь, если он хоть немного знаком с картами для гармина, увидев надпись JNX, сразу понимает, что как-то здесь можно сделать jnx-файл. При надписи Export это менее очевидно. Т.е. надо придумать, как сделать не сильно хуже, чем сейчас -- может быть можно сделать иконку на эту тему.

С иконкой экспериментировал, но ничего хорошего у меня не получилось. Кроме того нет понимания иконки с какими лицензиями можно использовать.

Пока есть идея только назвать кнопку "JNX+", а остальное - в тултипе.

 >   Превращение контекстного меню в диалог. Диалог надо делать отдельный, как в остальных контролах, туда же переносить список зумов.

Посмотрел на диалоги. Мне не нравится как они реализованы и их поведение. (правильно ли я понимаю что под диалогами понимается настройка/добавление слоя? Или же имелись ввиду сворачивающиеся контролы?)

По реализации:
 * нет общей части реализации диалогов
 * оба диалога которые есть сейчас - находятся в одном файле, html-шаблоны - в середине
 * по методам не понятно к какому из диалогов они относятся
 * knockout вроде используется, но очень ограниченно. Кажется что часть методов можно было бы в knockout модель переместить (хотя я невнимательно смотрел). Меня удивил кусок со смесью knockout-шаблонизации и js-шаблонизации. Переделал на knockout: https://github.com/tarwirdur/nakarte/commit/f8f4c1a11b03e4df936567447c3347d7a27508fe ( почему там так странно сделано? )
 
По поведению:

 * диалоги не являются модальными. То есть если сделать такой диалог для экспорта - можно будет при открытом диалоге нажать на "JNX" и селектор с выбранной областью пропадет. Конечно можно блокировать кнопку при открытии диалога, но возможно удобнее блокировать все кроме диалога.
 * нет закрытия диалога по escape, что особо не удобно при маленьком окне по высоте - кнопки уезжают под нижнюю границу экрана и нажать их нельзя.

Можно сделать сворачивающийся контрол, как у других элементов, но не хотелось бы им занимать много места на экране - впрочем можно сделать сворачивающиеся блоки, как в выборе размера страницы и отступов в контроле печати.

wladich commented:
При попытке сделать jnx или rmaps получаю в консоли
```
index.js:204 Uncaught TypeError: Cannot read property 'extension' of undefined
    at NewClass.startExport (index.js:204)
    at callback (index.js:176)
    at Contextmenu.onItemClick (index.js:278)
```

tarwirdur commented:
> index.js:204 Uncaught TypeError: Cannot read property 'extension' of undefined

Не понимаю, почему у меня при тестовом запуске работало. Возможно перед коммитом откатил случайно изменения.

Исправил.

wladich commented:
> Посмотрел на диалоги. Мне не нравится как они реализованы и их поведение. (правильно ли я понимаю что под диалогами понимается настройка/добавление слоя? Или же имелись ввиду сворачивающиеся контролы?)

Под "диалогом" я подразумевал контролы печати, треков и т.п., не настройки слоёв.

> * knockout вроде используется, но очень ограниченно. Кажется что часть методов можно было бы в knockout модель переместить (хотя я невнимательно смотрел). Меня удивил кусок со смесью knockout-шаблонизации и js-шаблонизации. Переделал на knockout: [tarwirdur@f8f4c1a](https://github.com/tarwirdur/nakarte/commit/f8f4c1a11b03e4df936567447c3347d7a27508fe) ( почему там так странно сделано? )

Да, так лучше, но не надо делать рефакторинг модулей, не относящихся к пул-реквесту. Если считаете, что рефакторинг нужен, надо сделать отдельный PR.

> 
> По поведению:
> 
> * диалоги не являются модальными. То есть если сделать такой диалог для экспорта - можно будет при открытом диалоге нажать на "JNX" и селектор с выбранной областью пропадет. Конечно можно блокировать кнопку при открытии диалога, но возможно удобнее блокировать все кроме диалога.

Не надо ничего блокировать. Если пользователь во время сборки pdf или jnx хочет работать дальше, в т.ч. менять область выделения, не надо ему мешать. Блокируется только запуск изготовления нового pdf/jnx. Кстати, хорошо бы сделать общую блокировку, но не обязательно в этом таске.

> Можно сделать сворачивающийся контрол, как у других элементов, но не хотелось бы им занимать много места на экране - впрочем можно сделать сворачивающиеся блоки, как в выборе размера страницы и отступов в контроле печати.

Данный конкретный контрол должен быть по-умолчанию свёрнут, как компас или координаты.

tarwirdur commented:
@wladich Просьба прокомментировать внешний вид и поведеление. Выложил текущее состояние сюда: http://nkrt.mikhe.ru:22837

код не готов, жду только feedback.

wladich commented:
@tarwirdur 
Код не смотрел, внешний вид и поведение, по-моему, получились отлично:
 - состояние в hashState -- хорошо
 - последний формат  в localStorage -- отличчно

Небольшие замечания:
1. Имя слоя попробовать перенести наверх, перед выбором формата, заодно убрать разделяющие полоски, они здесь не нужны
2. Зафиксировать ширину контрола, чтобы правая граница не дёргалась при изменении размера выделения.
3. Прогресс-бар отцентрировать.
4. Сделать  немного более подробнее тултип для красных строк (необязательно, могу потом доделать на свой вкус).
5. Строка "Zoom level", наверное избыточна.
6. Что-то сделать с оформлением ссылок. Синих ссылок сейчас в интерфейсе нигде нет, всё чёрно-серо-белое; бегающее подчеркивание, да ещё с разрывами на скобках, мне совсем не нравится. Сходу вижу варианты: a) точечное подчеркивание, типа как в надписи "100 m" на контроле печати, б) серый фон при наведении мыши как в текущем варианте в контекстном меню; может ещё что...
7. Про надпись на кнопке:  пока форматов всего два, думаю, можно сделать JNX/RMap в две строки. Потом, конечно, придётся что-то более осмысленное придумать, но пока можно и так сделать.

tarwirdur commented:
Сделал изменения в ui. Обновил стенд.

На выходных собираюсь:

1. блокировать ссылки при запущенном экспорте
2. почистить код
3. сделать rebase и разбить на несколько PR.

> Зафиксировать ширину контрола, чтобы правая граница не дёргалась при изменении размера выделения.

сделал `min-width: 300px`. Возможно, если тайлов очень много - может и дернуться, но это наверное лучше чем обрезать строку или делать многострочной опцию.

>  Прогресс-бар отцентрировать.

Отцентрировал. Лично мне внешне выравнивание по правому краю нравилось чуть больше, но в целом все равно.

>  Сделать немного более подробнее тултип для красных строк (необязательно, могу потом доделать на свой вкус).

Тут я плохо понимаю что писать в подсказку. Это оставлю.

>    Что-то сделать с оформлением ссылок.

Взял стили из contextmenu.

tarwirdur commented:
Пересоздал PR от новой ветки. Этот закрываю.

https://github.com/wladich/nakarte/pull/236
